sdk/program/src/message.rs

```diff
@@ -326,23 +347,31 @@ impl Message {
         self.program_position(i).is_some()
     }
 
-    pub fn is_writable(&self, i: usize) -> bool {
-        i < (self.header.num_required_signatures - self.header.num_readonly_signed_accounts)
+    pub fn is_writable(&self, i: usize, demote_sysvar_write_locks: bool) -> bool {
+        (i < (self.header.num_required_signatures - self.header.num_readonly_signed_accounts)
             as usize
             || (i >= self.header.num_required_signatures as usize
                 && i < self.account_keys.len()
-                    - self.header.num_readonly_unsigned_accounts as usize)
+                    - self.header.num_readonly_unsigned_accounts as usize))
+            && !{
+                let key = self.account_keys[i];
```

Comment:
I think this indexing is free of out-of-bound access, because `i` should be within `self.account_keys.len()` always.